### PR TITLE
BookStack integration

### PR DESCRIPTION
**Pull request by flifloo**
Add [BookStack](https://github.com/BookStackApp/BookStack) integration based on the Confluence integration

ToDo :
- [X] Base integration
- [x] Test pages Markdown content
- [x] Test API rate limits
- [x] Find a way to fetch user avatar thought ID
- [ ] Add page chapter in the location field

### CODE REVIEW

**Comment by Roey7** [comment]
Awsome dude! 
Are you on Discord? 
Let's chat.

**Comment by Roey7** [comment]
https://discord.gg/EJYfBkd4

File: app/data_sources/bookstack.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        data = r.json().get("data")
+        records = len(data)
+        if records >= count:
+            data += self.get_books(count, offset+records)
+        return data
+
+    def get_all_books(self) -> list[dict]:
+        offset = 0
+        count = 500
+        return self.get_books(count, offset)
```
Comment (Roey7) [inline app/data_sources/bookstack.py:50]: What if you have more than 500 books?

File: app/data_sources/bookstack.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    def request(self, method, url_path, *args, **kwargs):
+        url = urljoin(self.base_url, url_path)
+        return super().request(method, url, *args, **kwargs)
+
+    def get_books(self, count: int = 500, offset: int = 0):
+        r = self.get("/api/books", params={"count": count, "offset": offset, "sort": "+updated_at"},
+                     headers={"Content-Type": "application/json"})
+        data = r.json().get("data")
+        records = len(data)
+        if records >= count:
```
Comment (Roey7) [inline app/data_sources/bookstack.py:43]: `count=500` and you get only `130` books,  you still wanna add them right?

File: app/data_sources/bookstack.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        r = self.get("/api/pages", params={"count": count, "offset": offset, "sort": "+updated_at",
+                                           "filter[book_id]": book["id"]},
+                     headers={"Content-Type": "application/json"})
+        data = r.json().get("data")
+
+        for page in data:
+            page.update({"book": book})
+
+        records = len(data)
+        if records >= count:
```
Comment (Roey7) [inline app/data_sources/bookstack.py:62]: Same here. I think you don't need that if statement

File: app/data_sources/bookstack.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    def get_books(self, count: int = 500, offset: int = 0):
+        r = self.get("/api/books", params={"count": count, "offset": offset, "sort": "+updated_at"},
+                     headers={"Content-Type": "application/json"})
+        data = r.json().get("data")
+        records = len(data)
+        if records >= count:
+            data += self.get_books(count, offset+records)
+        return data
+
+    def get_all_books(self) -> list[dict]:
```
Comment (Roey7) [inline app/data_sources/bookstack.py:47]: Place replace `list` and `dict` to their typing `List`, `Dict` (we use python 3.8 rn)

File: app/data_sources/bookstack.py (hunk 2 above)
Comment (flifloo) [reply app/data_sources/bookstack.py:43]: If you get less than `count` you don't need to recursively fetch again with an increased `offset` ?

File: app/data_sources/bookstack.py (hunk 1 above)
Comment (flifloo) [reply app/data_sources/bookstack.py:50]: This is only the default page size per requests, when `count` is exceeded we start a recursive request

File: app/data_sources/bookstack.py (hunk 2 above)
Comment (Roey7) [reply app/data_sources/bookstack.py:43]: oh my bad, I see what you did there.
Can you write it in a non-recursive way tho? a `while has_more` or somth might be more readable. - thanks

File: app/data_sources/bookstack.py (hunk 4 above)
Comment (flifloo) [reply app/data_sources/bookstack.py]: Fixed by [ec6f696](https://github.com/GerevAI/gerev/pull/21/commits/ec6f696c4cd4a8f92bab84984684e6717ff955e7)

File: app/data_sources/bookstack.py (hunk 2 above)
Comment (flifloo) [reply app/data_sources/bookstack.py]: Done in [2140f9f](https://github.com/GerevAI/gerev/pull/21/commits/2140f9feeaa06d8e8ed694c4489e689439b271ef)

**Comment by flifloo** [comment]
Rabse to master + implement #25

**Pull request merged**